### PR TITLE
IGNITE-14667 Optimization of key criterion inference for aggregated watch.

### PR DESCRIPTION
**Pull request by kgusakov**

### CODE REVIEW

File: modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/MetaStorageManager.java [hunk 1]
```diff
@@ -180,11 +182,7 @@ public synchronized void deployWatches() {
         if (watch.isEmpty())
```
Comment (sanpwc) [inline modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/MetaStorageManager.java:182]: Formatting.

File: modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/watch/KeyCriterion.java [hunk 2]
```diff
@@ -17,62 +17,80 @@
...
-            this.key = key;
-        }
+    public static KeyCriterion union(KeyCriterion keyCriterion1, KeyCriterion keyCriterion2) {
+        Optional<KeyCriterion> res;
 
-        /** {@inheritDoc} */
-        @Override public IgniteBiTuple<Key, Key> toRange() {
-            return new IgniteBiTuple<>(key, key);
-        }
+        if (keyCriterion1.getHierarchyOrder() < keyCriterion2.getHierarchyOrder())
```
Comment (sanpwc) [inline modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/watch/KeyCriterion.java:60]: To be honest I don't like an idea of union ordering. Let's make it associative.

File: modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/watch/KeyCriterion.java [hunk 3]
```diff
@@ -17,62 +17,80 @@
...
+            res = keyCriterion1.doUnion(keyCriterion2);
 
-        /** {@inheritDoc} */
-        @Override public boolean contains(Key key) {
-            return this.key.equals(key);
-        }
+        if (res.isPresent())
+            return res.get();
+        else
+            throw new UnsupportedOperationException("Can't calculate the union of " + keyCriterion1.getClass() + " and " +
```
Comment (sanpwc) [inline modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/watch/KeyCriterion.java:68]: Why it's possible?

File: modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/watch/KeyCriterion.java [hunk 4]
```diff
@@ -91,20 +109,106 @@ public RangeCriterion(Key from, Key to) {
...
+            else if (keyCriterion instanceof CollectionCriterion) {
+                from = Collections.min(((CollectionCriterion)keyCriterion).keys);
+                to = nextKey(Collections.max(((CollectionCriterion)keyCriterion).keys));
+            }
+            else if (keyCriterion instanceof RangeCriterion) {
+                from = ((RangeCriterion)keyCriterion).from;
+                to = ((RangeCriterion)keyCriterion).to;
+            }
+            else
+                return Optional.empty();
```
Comment (sanpwc) [inline modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/watch/KeyCriterion.java:145]: As was mentioned above
> be honest I don't like an idea of union ordering. Let's make it associative.
I think we should rewrite doUnion() logic in order not to return empty result.

File: modules/metastorage/src/test/java/org/apache/ignite/internal/metastorage/WatchAggregatorTest.java [hunk 5]
```diff
@@ -113,6 +115,70 @@ public void testCancelByFalseFromListener() {
 
     }
 
+    @Test
```
Comment (sanpwc) [inline modules/metastorage/src/test/java/org/apache/ignite/internal/metastorage/WatchAggregatorTest.java:118]: javadoc

File: modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/watch/KeyCriterion.java [hunk 6]
```diff
@@ -118,45 +222,113 @@ public CollectionCriterion(Collection<Key> keys) {
...
-            return new IgniteBiTuple<>(Collections.min(keys), Collections.max(keys));
+        @Override public boolean contains(Key key) {
+            return keys.contains(key);
         }
 
         /** {@inheritDoc} */
-        @Override public boolean contains(Key key) {
-            return keys.contains(key);
+        @Override protected Optional<KeyCriterion> doUnion(KeyCriterion keyCriterion) {
+            var newKeys = new ArrayList<>(keys);
```
Comment (sanpwc) [inline modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/watch/KeyCriterion.java:231]: Hmm, why not to use Set? Am i right that in current solution it's possible to duplicate keys?

File: modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/watch/KeyCriterion.java [hunk 7]
```diff
@@ -118,45 +222,113 @@ public CollectionCriterion(Collection<Key> keys) {
...
-            if (bytes[bytes.length - 1] != Byte.MAX_VALUE)
-                bytes[bytes.length - 1]++;
+        @Override public boolean contains(Key key) {
+            return this.key.equals(key);
+        }
+
+        /** {@inheritDoc} */
+        @Override protected Optional<KeyCriterion> doUnion(KeyCriterion keyCriterion) {
+            if (keyCriterion instanceof ExactCriterion)
+                return Optional.of(new CollectionCriterion(Arrays.asList(key, ((ExactCriterion)keyCriterion).key)));
```
Comment (sanpwc) [inline modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/watch/KeyCriterion.java:300]: Same, as above, are we going to create Collection Criterion for two exact criterions that equals one another?

File: modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/watch/KeyCriterion.java [hunk 8]
```diff
@@ -91,72 +93,216 @@ public RangeCriterion(Key from, Key to) {
...
+                from = ((RangeCriterion)keyCriterion).from;
+                to = ((RangeCriterion)keyCriterion).to;
+            }
+            else if (!swapTry)
+                return keyCriterion.union(this, true);
+            else throw KeyCriterion.unsupportedUnionException(this, keyCriterion);
+
+            return new RangeCriterion(
+                Collections.min(Arrays.asList(this.from, from)),
+                Collections.max(Arrays.asList(this.to, to))
```
Comment (sanpwc) [inline modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/watch/KeyCriterion.java:134]: Seems that we'll get NPE if  to is null
```
     * @param keyTo End key of range (exclusive). Could be {@code null}.
     * @return Cursor built upon entries corresponding to the given range and revision.
     * @throws OperationTimeoutException If the operation is timed out.
     * @throws CompactedException If the desired revisions are removed from the storage due to a compaction.
     * @see ByteArray
     * @see Entry
     */
    @NotNull
    Cursor<Entry> range(@NotNull ByteArray keyFrom, @Nullable ByteArray keyTo);
```

File: modules/metastorage/src/test/java/org/apache/ignite/internal/metastorage/WatchAggregatorTest.java [hunk 9]
```diff
@@ -113,6 +128,122 @@ public void testCancelByFalseFromListener() {
...
+
+        var keyCriterion = watchAggregator.watch(0, null).get().keyCriterion();
+        var expKeyCriterion = new KeyCriterion.RangeCriterion(
+            new Key("key0"), new Key("key4"));
+        assertEquals(expKeyCriterion, keyCriterion);
+    }
+
+    /**
+     *
+     */
```
[the remark this answers is not in the document]
Comment (kgusakov) [reply modules/metastorage/src/test/java/org/apache/ignite/internal/metastorage/WatchAggregatorTest.java:278]: Added `testThatKeyCriteriaUnionAssociative`

File: modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/watch/KeyCriterion.java [hunk 10]
```diff
@@ -91,74 +93,250 @@ public RangeCriterion(ByteArray from, ByteArray to) {
...
+            }
+            else if (keyCriterion instanceof RangeCriterion) {
+                from = ((RangeCriterion)keyCriterion).from;
+                to = ((RangeCriterion)keyCriterion).to;
+            }
+            else if (!swapTry)
+                return keyCriterion.union(this, true);
+            else throw KeyCriterion.unsupportedUnionException(this, keyCriterion);
+
+            return new RangeCriterion(
```
Comment (sk0x50) [inline modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/watch/KeyCriterion.java:132]: Looks like this line is not reachable.

File: modules/metastorage/src/test/java/org/apache/ignite/internal/metastorage/WatchAggregatorTest.java [hunk 11]
```diff
@@ -29,15 +32,22 @@
...
+
 import static org.junit.jupiter.api.Assertions.assertEquals;
 import static org.mockito.ArgumentMatchers.any;
 import static org.mockito.Mockito.mock;
 import static org.mockito.Mockito.times;
 import static org.mockito.Mockito.verify;
 import static org.mockito.Mockito.when;
 
+/**
+ *
```
Comment (sk0x50) [inline modules/metastorage/src/test/java/org/apache/ignite/internal/metastorage/WatchAggregatorTest.java:44]: Empty Javadoc looks very sad. :(

File: modules/metastorage/src/test/java/org/apache/ignite/internal/metastorage/WatchAggregatorTest.java [hunk 12]
```diff
@@ -131,6 +147,167 @@ public void testCancelByFalseFromListener() {
...
+     *
+     */
+    @Test
+    public void testOneCriterionInference() {
+        var watchAggregator = new WatchAggregator();
+
+        watchAggregator.add(new ByteArray("key"), null);
+
+        var keyCriterion = watchAggregator.watch(0, null).get().keyCriterion();
+        assertEquals(new KeyCriterion.ExactCriterion(new ByteArray("key")), keyCriterion);
```
Comment (sk0x50) [inline modules/metastorage/src/test/java/org/apache/ignite/internal/metastorage/WatchAggregatorTest.java:160]: Please try to add a human-readable message that provides details.
"Test failed: expected 1, actual 2" looks very concise to me.

File: modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/watch/KeyCriterion.java (hunk 10 above)
Comment (sk0x50) [reply modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/watch/KeyCriterion.java:132]: My fault. This line is obviously reachable :)

**Pull request closed without merging**